sparkly/writer.py

```diff
@@ -255,6 +275,188 @@ def write_partition_to_kafka(messages):
+            raise ValueError('redis: max pipeline size must be positive')
+
+        if mode not in {'overwrite', 'ignore', 'flush'}:
+            raise ValueError(
+                'redis: only overwrite (default), ignore and flush modes are supported.'
+            )
+
+        key_name = '_sparkly_redis_key_col_{}'.format(uuid.uuid4().hex)
+        value_name = '_sparkly_redis_value_col_{}'.format(uuid.uuid4().hex)
+
+        # Compute the key
+        df = self._df.withColumn(
+            key_name,
+            F.concat_ws(
+                key_delimiter,
+                *(
+                    ([F.lit(key_prefix).astype(T.StringType())] if key_prefix else []) +
+                    [F.col(col_name).astype(T.StringType()) for col_name in key_by]
+                )
+            )
+        )
+
+        if exclude_key_columns:
+            for col_name in key_by:
+                df = df.drop(col_name)
+
+        if group_by_key:
+            df = (
+                df
+                .withColumn(
+                    value_name,
+                    F.struct(*[col for col in df.columns if col != key_name])
+                )
+                .groupBy(key_name)
+                .agg(F.expr('collect_list( {} )'.format(value_name)).alias(value_name))
+            )
+
+        # Repartition if needed to achieve specified parallelism
+        df = df.coalesce(parallelism or df.rdd.getNumPartitions())
+
+        def compress(data, protocol=None):
+            if protocol is None:
+                return data
+            elif protocol == 'bzip2':
+                return bz2.compress(data)
+            elif protocol == 'zlib':
+                return zlib.compress(data)
+            elif protocol == 'gzip':
+                try:
+                    return gzip.compress(data)
+                except AttributeError:  # py27
+                    compressor = zlib.compressobj(
+                        zlib.Z_DEFAULT_COMPRESSION,
+                        zlib.DEFLATED,
+                        zlib.MAX_WBITS | 16,
+                    )
+                    return compressor.compress(data) + compressor.flush()
+            else:
+                raise ValueError('unknown compression protocol {}'.format(protocol))
+
+        # Define the function that will write each partition to redis
+        def _write_redis(partition, redis, expire, compress, max_pipeline_size, mode):
+            pipeline = redis().pipeline(transaction=False)
+            add_to_pipeline = partial(pipeline.set, ex=expire, nx=(mode == 'ignore'))
+
+            for rows in chunks(partition, max_pipeline_size):
+                for row in rows:
+                    row = row.asDict(recursive=True)
+
+                    key = row.pop(key_name)
+                    if group_by_key:
+                        row = row[value_name]
+                    if exclude_null_fields:
+                        for null_field in list(f for f in row.keys() if row[f] is None):
```

Comment:
what is the reason to transform to list here?
`for null_field in (key for key in row.items() if value is None):` does the same but in more efficient way.